CGIV-5637: Reconnect on connection error

RichardHeelin opened the pull request.

Commit:
CGIV-5637: Reuse a single connection to redis so that we don't get errors with trying to connect once we have started processing the cron

Commit:
CGIV-5637: In the event that we error, stop processing and return error

Commit:
CGIV-5637: Correctly handle multi-line output, previously was only reading the first line

Commit:
CGIV-5637: INFO is an expensive operation so lets use PING to verify we are talking to redis

Commit:
CGIV-5637: Reconnect to redis if there is an issue with the connection after we have connected

Commit:
CGIV-5637: If connection drops, attempt to reconnect upto 5 times with an incremental backoff

Commit:
CGIV-5637: No wait for initial reconnect as we don't know when we lost the connection

Commit:
CGIV-5637: Added a new option, CRONLOCK_RECONNECT_BACKOFF, which controls the reconnect backoff

Commit:
CGIV-5637: Make sure we hide any errors on connection test

Commit:
CGIV-5637: Only fatal on initial connection

Commit:
CGIV-5637: If the value is not set, for instance redis has been flushed, the length returned is -1

nickpeirson commented on cronlock:
```diff
@@ -59,8 +60,26 @@ function get_master {
+function reconnect {
+  attempts=0
+  until cmd "PING" 0 2> /dev/null | grep 'PONG' > /dev/null; do
+    if [ $attempts -ge 5 ]; then
```
Parametrise reconnect attempts

nickpeirson commented on cronlock:
```diff
@@ -59,8 +60,26 @@ function get_master {
+
 function cmd {
-  exec 3<>/dev/tcp/${CRONLOCK_HOST}/${CRONLOCK_PORT} || fatal "Redis down"
+  [[ -z "$2" || "$2" -gt 0 ]] && reconnect >&2
```
Split out the underlying command handling to be used from `reconnect`

nickpeirson commented on cronlock:
```diff
@@ -59,8 +60,26 @@ function get_master {
+
+function reconnect {
+  attempts=0
+  until cmd "PING" 0 2> /dev/null | grep 'PONG' > /dev/null; do
```
Try to catch errors rather than pre-empt

Commit:
CGIV-5637: Refactored the reconnect method so that we no longer have to make two redis calls per cmd

Commit:
CGIV-5637: Limit reconnect to connection errors rather than all errors

Commit:
CGIV-5637: Added an explanation for the >= 0 check

Only occurs if we do a GETSET after redis has been restarted or flushed resulting the the key not been set

Commit:
CGIV-5637: Allow users to customise reconnect and timeout arguments

Commit:
CGIV-5637: Allow users to customise reconnect and timeout arguments

Commit:
CGIV-5637: Removed error state enforcement as it breaks the timeout handler

Commit:
CGIV-5637: Renamed timeout to better describe it's use

nickpeirson commented:
:+1:
